### PR TITLE
Allow to use subfolders of src in the build of john

### PR DESCRIPTION
**Pull request by ghost**
Tested using the latest code (2023-05-02 13:00 UTC):
- [x] Android ARM
- [x] Android X86
- [x] FreeBSD
- [x] Linux
- [x] MacOS ARM
- [x] MacOS X86
- [x] OpenCL
- [x] Solaris
- [x] Windows cygwin
- [x] Windows Wine

To simplify the understanding and review process, patches should be:
- Commit 1: `git mv`, 250 files, does not build (nothing to review);
- Commit 2: `update includes`, 350 files, does not build (glance at it);
- Commit 3: `update configure and makefile`, 3 files, compiles correctly (deserves to be reviewed);

### CODE REVIEW

File: src/Makefile.in [hunk 1]
```diff
@@ -855,6 +855,9 @@ clean:
 	@for exe in ${PROJ}; do \
 	  ($(RM) $$exe.exe) \
 	done
+	@for object in ${JOHN_OBJS}; do \
+	  ($(RM) $$object) \
+	done
```
Comment (claudioandre-br) [inline src/Makefile.in:860]: This is probably wrong. I'm not seeing what I expect to see during `make clean`.

**Comment by ghost** [comment]
I converted to a draft PR.

To make the CI happy, I had to disable `-I` as seen below. I cannot understand how this can be a bad thing.

```diff
   LDFLAGS="-fsanitize=undefined -fsanitize-undefined-trap-on-error $LDFLAGS"
fi
# As the last thing to do before building, add the 'include' folder.
-CPPFLAGS="$CPPFLAGS -I$(pwd)"
+# CPPFLAGS="$CPPFLAGS -I$(pwd)"

if test "x$endian" = "xbig"; then
   CFLAGS="$CFLAGS -DARCH_LITTLE_ENDIAN=0"
```

New gcc compilers deal with it, old ones don't. Anyone have a hint? Is it any kind of x-rated C?

Without the '-I' we'll need to change all moved files to include files like "../jumbo.h"

**Comment by ghost** [comment]
Replacing `-I` with `-iquote` makes gcc happy and makes sense. Let's go ahead and add more tests.

**Comment by claudioandre-br** [comment]
I added a new commit that moves `nt*` to plugins and bitlocker to `wip` (names are subject to change). Let's hear the CI.

**Comment by magnumripper** [comment]
There's also compatibility with BSD `make` to consider IIRC.

Regarding directory structure, some have an `include/` directory for all headers (in our case maybe except core files).

**Comment by claudioandre-br** [comment]
- My CI is going to test it on FreeBSD (by the end of the day). 
- I'll move something to `include/`.

Instead of `wip`, we can use `community` or `jumbo` or ... ??

**Comment by magnumripper** [comment]
Just brainstorming. This example does not put headers away from their corresponding source, and does not separate plugins from non-plugins.

`src/` (core files including jumbo's core, eg. loader.[ch], jumbo.[ch])
`src/modes/` (modes, eg. inc.[ch], subsets.[ch])
`src/opencl` (eg. opencl_common.c, opencl_autotune.h)
`src/ztex` (already exists - ztex core stuff)
`src/formats` (eg. md5crypt_common.c, md5crypt_common.h)
`src/formats/cpu` (eg. md5crypt_long_fmt.c)
`src/formats/opencl` (eg. opencl_md5crypt_fmt_plug.c)
`src/formats/ztex` (eg. ztex_md5crypt.c)
(...)

**Comment by ghost** [comment]
I tried using these folders.
```
modes
formats
plugins
opencl
```

They are a good starting point, but we should focus on the plugins now or the patch will get big.

**Comment by ghost** [comment]
The `plugin_deps.pl` is **not** 100%, but perl is really odd. What the `while` is doing? A buit in `sed`?
***
I can move more plugins into the `plugins/` subdirectory, but I'd rather stop experimenting on this PR. Too many moving parts.

**Comment by magnumripper** [comment]
> The `plugin_deps.pl` is **not** 100%, but perl is really odd. What the `while` is doing? A buit in `sed`?

The `while (<$fh>)` iterates over lines from the filehandle (until EOF) and `if (/^\s*#\s*include\s+"([^"]+)"/)` is more like a grep but it ends up with the match for `([^"]+)` as `$1`. The inner `while` does a `sed` but with perlre powers. Then just in case anyone would still be able to follow, the function recurses 😵 I believe I wrote it but it'd take me quite a while figuring it out enough to change anything.

**Comment by ghost** [comment]
CI is happy (Ubuntu 16, OpenCL, Fedora old, locally I am at Ubuntu 23.04).

Should I run:
```
$ git mv $(ls *_fmt_plug.[c-h] | grep -Ev opencl) plugins/
```

Amend and test again? 

After that, my own CI will take place.

**Comment by magnumripper** [comment]
Sure, let's test it. I think we should await @solardiz before merging anything though. And we need to deal with the legacy Makefile.

**Comment by solardiz** [comment]
> Replacing `-I` with `-iquote` makes gcc happy and makes sense.

How portable is `-iquote` to other compilers? I think I'd rather have us update the include paths inside source files.

**Comment by ghost** [comment]
Force-pushed to document a [regression fix](https://github.com/openwall/john/compare/460338b2514a964cee7833bbb4d823ae1b68b879..4d4ee22bc3335b57b6c9c30ce317e0313a0aad0a)

> How portable is `-iquote` to other compilers? I think I'd rather have us update the include paths inside source files.

All include paths are updated now. I created a script to automate this, it is saved in unused.

**Comment by ghost** [comment]
Testing all CPU plugins inside `plugins/`.

**Comment by ghost** [comment]
There was some bashi-ism in the PR, I fixed it and now it's passing on FreeBSD and Mac M1. 

I'll do a full CI rebuild [1] tomorrow and we'll keep the PR open for experimentation.

[1] Since I disabled the zip format, changed some parts of the code, disabled `werror` and so on.

**Comment by claudioandre-br** [comment]
I updated all the include paths inside source files and removed all `--iquotes`. 

I created a script to automate this, it is saved in unused.

**Comment by solardiz** [comment]
Thank you guys for starting to experiment with this. I'm afraid I won't get to it earlier than in 2+ months from now. When I do, I'd prefer us to have proposed changes in a form that's potentially ready to merge - separate commits fully implementing each change, one at a time. Of course, chances are we'll end up revising them rather than merging them as-is, but we would be much closer to merging than we're here with a confusing mix of what clearly are unfinished experiments not to be merged and changes that might be proposed for merging.

**Pull request closed without merging**